File: 06-DevOps/05-IaC与配置管理/Ansible集合与依赖解析/anscoll.go
// Ansible Collections 的 FQCN 解析、collections 搜索路径作用域与 runtime.yml 元数据。
// 口径与 Python 版一致，来自实际读过的官方原文（见 README 参考资料）。
package main

import (
	"strconv"
	"strings"
)

// 官方："an FQCN is still required for non-action or module plugins
// (for example, lookups, filters, and tests)"
var nonSearchable = map[string]bool{
	"lookup": true, "filter": true, "test": true,
	"connection": true, "callback": true, "vars": true, "cache": true,
}

type FQCN struct {
	Namespace  string
	Collection string
	Rest       []string
}

// ParseFQCN 返回 nil 表示不是 FQCN。
func ParseFQCN(ref string) *FQCN {
	parts := strings.Split(ref, ".")
	if len(parts) < 3 {
		return nil
	}
	return &FQCN{parts[0], parts[1], parts[2:]}
}

// IsValidPlaybookName：官方说集合里的 playbook 名不允许出现连字符。
func IsValidPlaybookName(name string) bool {
	return name != "" && !strings.Contains(name, "-")
}

// ModuleUtilsImport：from ansible_collections.{ns}.{coll}.plugins.module_utils.{util} import ...
func ModuleUtilsImport(namespace, collection, util string) string {
	return "ansible_collections." + namespace + "." + collection + ".plugins.module_utils." + util
}

// ResolvePlugin 按 collections 列表顺序找短名插件，返回 (fqcn, reason)。
// reason ∈ fqcn / search_path / requires_fqcn / not_found
func ResolvePlugin(shortName string, collections []string, available map[string]bool,
	ptype string) (string, string) {
	if ParseFQCN(shortName) != nil {
		if available[shortName] {
			return shortName, "fqcn"
		}
		return "", "not_found"
	}
	if nonSearchable[ptype] {
		return "", "requires_fqcn"
	}
	for _, coll := range collections {
		cand := coll + "." + shortName
		if available[cand] {
			return cand, "search_path"
		}
	}
	return "", "not_found"
}

// ResolveInRole：角色内只看角色自己的 collections，不继承 playbook 的。
func ResolveInRole(shortName string, playbookCollections, roleCollections []string,
	available map[string]bool, ptype string) (string, string) {
	return ResolvePlugin(shortName, roleCollections, available, ptype)
}

// releaseTuple 截掉预发布段（官方：Ansible 会 truncate prerelease segments）。
func releaseTuple(v string) [3]int {
	core := strings.Split(v, "+")[0]
	seps := []string{"a", "b", "rc", "dev", ".post"}
	best := -1
	for _, s := range seps {
		if i := strings.Index(core[1:], s); i >= 0 && (best < 0 || i+1 < best) {
			best = i + 1
		}
	}
	if best > 0 {
		core = core[:best]
	}
	core = strings.TrimRight(core, ".")
	var out [3]int
	i := 0
	for _, p := range strings.Split(core, ".") {
		if i >= 3 {
			break
		}
		digits := ""
		for _, ch := range p {
			if ch >= '0' && ch <= '9' {
				digits += string(ch)
			}
		}
		n := 0
		if digits != "" {
			n, _ = strconv.Atoi(digits)
		}
		out[i] = n
		i++
	}
	return out
}

func cmpTuple(a, b [3]int) int {
	for i := 0; i < 3; i++ {
		if a[i] < b[i] {
			return -1
		}
		if a[i] > b[i] {
			return 1
		}
	}
	return 0
}

// RequiresAnsibleOK：meta/runtime.yml 的 requires_ansible，PEP440 说明符，逗号分隔。
// 官方：Ansible 2.11.0b1 与 requires_ansible ">=2.11" 兼容（预发布段被截断）。
func RequiresAnsibleOK(spec, ansibleVersion string) (bool, error) {
	ver := releaseTuple(ansibleVersion)
	for _, raw := range strings.Split(spec, ",") {
		c := strings.TrimSpace(raw)
		if c == "" {
			continue
		}
		matched := false
		for _, op := range []string{">=", "<=", "!=", "=", ">", "<"} {
			if !strings.HasPrefix(c, op) {
				continue
			}
			want := releaseTuple(strings.TrimSpace(c[len(op):]))
			r := cmpTuple(ver, want)
			var ok bool
			switch op {
			case ">=":
				ok = r >= 0
			case "<=":
				ok = r <= 0
			case "=":
				ok = r == 0
			case "!=":
				ok = r != 0
			case ">":
				ok = r > 0
			case "<":
				ok = r < 0
			}
			if !ok {
				return false, nil
			}
			matched = true
			break
		}
		if !matched {
			return false, errUnrecognized(c)
		}
	}
	return true, nil
}

func errUnrecognized(c string) error {
	return &specErr{c}
}

type specErr struct{ c string }

func (e *specErr) Error() string { return "无法识别的说明符: " + e.c }

type Tombstone struct {
	RemovalVersion string
	WarningText    string
}

type RoutingEntry struct {
	Redirect    string
	Deprecation string
	Tombstone   *Tombstone
}

type RouteResult struct {
	Target   string
	Warnings []string
	Fatal    string
}

// RoutePlugin 应用 plugin_routing：redirect / deprecation / tombstone。
func RoutePlugin(routing map[string]map[string]RoutingEntry, ptype, pluginName string) RouteResult {
	r, ok := routing[ptype][pluginName]
	if !ok {
		return RouteResult{pluginName, nil, ""}
	}
	var res RouteResult
	if r.Tombstone != nil {
		res.Fatal = pluginName + " 已被移除（removal_version=" + r.Tombstone.RemovalVersion +
			"）：" + r.Tombstone.WarningText
		return res
	}
	if r.Deprecation != "" {
		res.Warnings = append(res.Warnings, r.Deprecation)
	}
	if r.Redirect != "" {
		res.Target = r.Redirect
		return res
	}
	res.Target = pluginName
	return res
}

File: 06-DevOps/05-IaC与配置管理/Ansible集合与依赖解析/anscoll.py
# -*- coding: utf-8 -*-
"""Ansible Collections 的 FQCN 解析、collections 搜索路径作用域与 runtime.yml 元数据。

口径全部来自实际读过的官方原文（见 README 参考资料）：
  - Developing collections / Collection structure：galaxy.yml 定义命名空间、module_utils 导入约定、
    meta/runtime.yml 的 requires_ansible 与 plugin_routing
  - Using collections in a playbook：`collections:` 是**有序搜索路径**、只作用于 action/module 与角色引用、
    **角色不继承 playbook 的 collections**
  - Galaxy user guide：requirements.yml 里声明集合依赖
"""
from __future__ import annotations

# 官方原文："an FQCN is still required for non-action or module plugins
# (for example, lookups, filters, and tests)"
NON_SEARCHABLE_TYPES = {"lookup", "filter", "test", "connection", "callback", "vars", "cache"}


class Routing:
    OFF = "none"
    REDIRECT = "redirect"
    DEPRECATED = "deprecation"
    TOMBSTONE = "tombstone"


def parse_fqcn(ref: str):
    """拆成 (namespace, collection, 剩余路径列表)。非 FQCN 返回 None。"""
    parts = ref.split(".")
    if len(parts) < 3:
        return None
    return parts[0], parts[1], parts[2:]


def is_valid_playbook_name(name: str) -> bool:
    """官方原文：集合里的 playbook 名**不允许**出现连字符。"""
    return "-" not in name and name != ""


def module_utils_import(namespace: str, collection: str, util: str) -> str:
    """官方约定：from ansible_collections.{ns}.{coll}.plugins.module_utils.{util} import ..."""
    return "ansible_collections.%s.%s.plugins.module_utils.%s" % (namespace, collection, util)


def resolve_plugin(short_name: str, collections, available, ptype="module"):
    """按 `collections:` 列表的顺序在可用集合里找短名插件。

    返回 (fqcn, reason)。两条官方约束：
      1) 官方："The `collections` keyword merely creates an ordered 'search path' for
         non-namespaced plugin and role references."
      2) 非 action/module 类型**即使**在搜索路径里也必须用 FQCN。
    """
    if parse_fqcn(short_name) is not None:
        return (short_name, "fqcn") if short_name in available else (None, "not_found")
    if ptype in NON_SEARCHABLE_TYPES:
        return None, "requires_fqcn"
    for coll in collections:
        candidate = "%s.%s" % (coll, short_name)
        if candidate in available:
            return candidate, "search_path"
    return None, "not_found"


def resolve_in_role(short_name, playbook_collections, role_collections, available,
                    ptype="module"):
    """官方："any roles you call in your playbook define their own collections search
    order; they do not inherit the calling playbook's settings. This is true even if
    the role does not define its own `collections`."

    因此角色内解析**只看角色自己的列表**，playbook 的列表不参与。
    """
    return resolve_plugin(short_name, role_collections, available, ptype)


def _release_tuple(v: str):
    """截掉预发布段（官方：Ansible 会 truncate prerelease segments）。"""
    core = v.split("+")[0]
    for sep in ("a", "b", "rc", "dev", ".post"):
        idx = core.find(sep, 1)
        if idx > 0:
            core = core[:idx]
            break
    core = core.rstrip(".")
    out = []
    for p in core.split("."):
        digits = "".join(ch for ch in p if ch.isdigit())
        out.append(int(digits) if digits else 0)
    while len(out) < 3:
        out.append(0)
    return tuple(out[:3])


def _cmp(a, b):
    return (a > b) - (a < b)


def requires_ansible_ok(spec: str, ansible_version: str) -> bool:
    """meta/runtime.yml 的 requires_ansible：PEP440 说明符，逗号分隔。

    官方原文："although the version is a PEP440 Version Specifier under the hood, Ansible
    deviates from PEP440 behavior by truncating prerelease segments from the Ansible
    version. This means that Ansible 2.11.0b1 is compatible with something that
    requires_ansible: '>=2.11'."
    """
    ver = _release_tuple(ansible_version)
    for raw in [c for c in spec.split(",") if c.strip()]:
        c = raw.strip()
        for op in (">=", "<=", "!=", "=", ">", "<"):
            if c.startswith(op):
                want = _release_tuple(c[len(op):].strip())
                r = _cmp(ver, want)
                ok = {">=": r >= 0, "<=": r <= 0, "=": r == 0, "!=": r != 0,
                      ">": r > 0, "<": r < 0}[op]
                if not ok:
                    return False
                break
        else:
            raise ValueError("无法识别的说明符: %r" % c)
    return True


def route_plugin(runtime_yml: dict, ptype: str, plugin_name: str):
    """应用 plugin_routing，返回 (实际目标, warnings, fatal)。

    官方三种动作：
      - redirect：改从另一个位置加载
      - deprecation：给自定义警告 + removal version/date（改名/挪位时通常同时给 redirect）
      - tombstone：整个移除，致命错误 + removal_version + warning_text
    """
    routing = runtime_yml.get("plugin_routing", {}).get(ptype, {}).get(plugin_name)
    warnings, fatal = [], None
    if not routing:
        return plugin_name, warnings, fatal
    if "tombstone" in routing:
        t = routing["tombstone"]
        fatal = "%s 已被移除（removal_version=%s）：%s" % (
            plugin_name, t.get("removal_version"), t.get("warning_text", ""))
        return None, warnings, fatal
    if "deprecation" in routing:
        warnings.append(routing["deprecation"].get("warning_text", ""))
    if "redirect" in routing:
        return routing["redirect"], warnings, fatal
    return plugin_name, warnings, fatal
